### PR TITLE
Use typed field mask paths

### LINKED ISSUE #598: [Refactor]: Use typed field mask paths
### Is there an existing issue for this?

- [x] I have searched the existing issues

### What do you think needs to change? Please describe.

Currently, the field mask paths are just strings. A typo or the wrong path is easy to overlook. If a new path is used, there are several places where new code needs to be added.

### Why should this be refactored?

Use an enum class for those paths. There they are typed, and we can make use of `when (path) { ... }` to handle all cases. We also can iterate over all paths for tests.

### What outcome do you expect?

No strings for DTO fields anymore. Conversion at the gRPC layer.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Slartibartfass2**
Closes #598

## What I have made

I recommend a review commit-by-commit.

This makes the field mask paths type safe and overall the project more robust and maintainable.

The branch currently includes #597 

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

### Author

- [x] I have updated the documentation accordingly and commented my code
  - ~~[ ] I have updated `AGENTS.md` if the changes affect project structure, commands, tooling, or conventions~~
- [x] I have manually tested my changes
- [x] I have added tests that prove my fix is effective or that my feature works

### Reviewer

- [x] I have checked the changes against the requirements

### CODE REVIEW

File: src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt [hunk 1]
```diff
@@ -124,18 +125,19 @@ class CriterionTableRepo(
...
-        CriterionTable.updateByIdAndGet(request.criterionId, ResultRow::toCriterion) {
-            for (field in paths) {
-                when (field) {
-                    "criterion.tag" -> it[tag] = request.tag
-                    "criterion.name" -> it[name] = request.name
-                    "criterion.description" -> it[description] = request.description
-                    "criterion.category" -> it[category] = request.category
+    override suspend fun updateCriterion(request: UpdateCriterionRequest, fields: Set<CriterionField>): Criterion =
+        db.query {
+            CriterionTable.updateByIdAndGet(request.criterionId, ResultRow::toCriterion) {
```
Comment (mowi12) [inline src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt:134]: Here, the empty-fields fast exit is missing:

```suggestion
        db.query {
            if (fields.isEmpty()) {
                return@query getCriterionById(request.criterionId).getOrThrow()
            }

            CriterionTable.updateByIdAndGet(request.criterionId, ResultRow::toCriterion) {
```

File: src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt [hunk 2]
```diff
@@ -114,3 +121,16 @@ fun userStatusFromGrpc(status: UserOuterClass.UserStatus): UserStatus = when (st
...
+fun projectFieldFromGrpc(fieldMaskPath: String): ProjectField =
+    ProjectField.entries.first { it.grpcPath == fieldMaskPath }
+
+fun criterionFieldFromGrpc(fieldMaskPath: String): CriterionField =
+    CriterionField.entries.first { it.grpcPath == fieldMaskPath }
+
+fun userFieldFromGrpc(fieldMaskPath: String): UserField = UserField.entries.first { it.grpcPath == fieldMaskPath }
+
+fun projectInfoFieldFromGrpc(fieldMaskPath: String): ProjectInfoField =
+    ProjectInfoField.entries.first { it.grpcPath == fieldMaskPath }
```
Comment (mowi12) [inline src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt]: This is duplicated code, let's improve this by rather using a more generic method:

```suggestion
/**
 * Converts the paths of the given field [mask] into the fields of the receiver, e.g. the entries of [UserField].
 *
 * The mask is normalized first, so duplicate paths are removed and paths that are covered by another path are folded
 * into it.
 *
 * @param T The type of the fields, e.g. [UserField].
 * @param mask The field mask of the request.
 * @param pathOf The field mask path of a field, e.g. [UserField.grpcPath].
 * @return The set of fields that the [mask] selects.
 * @throws NoSuchElementException If a path of the [mask] does not belong to any field of the receiver.
 */
fun <T> Collection<T>.fieldsFromGrpc(mask: FieldMask, pathOf: (T) -> String): Set<T> =
    FieldMaskUtil.normalize(mask).pathsList.mapTo(mutableSetOf()) { path -> first { pathOf(it) == path } }
```

This will now also remove the `TooManyFunctions` annotation.

You will need to update all call sites.

Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt]: This will now also remove the `TooManyFunctions` annotation.

File: src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt [hunk 3]
```diff
@@ -284,27 +271,26 @@ class CriterionTableRepoTest : RepositoryTest(arrayOf(CriterionTable, ProjectTab
...
+            // Excluded
+            for (excludedField in excluded) {
+                when (excludedField) {
+                    CriterionField.TAG -> assertEquals("Test Tag", updatedCriterion.tag)
+                    CriterionField.NAME -> assertEquals("Test Criterion", updatedCriterion.name)
+                    CriterionField.DESCRIPTION -> assertEquals("Test Description", updatedCriterion.description)
+                    CriterionField.CATEGORY -> assertEquals(CriterionCategory.EXCLUSION, updatedCriterion.category)
+                }
             }
         }
```
Comment (mowi12) [inline src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:295]: Let's add this test for the missing empty-field fast exit:

```kotlin
@Test
        fun `When a criterion is updated without any specified fields, then nothing is updated`() = runTest {
            val projectId = insertProjectAndGetId(createdBy = testUserId)
            val criterionId = insertCriterionAndGetId(projectId = projectId, createdBy = testUserId)
            val criterion = repo.getCriterionById(criterionId).getOrThrow()
            val request = UpdateCriterionRequest(
                criterionId = criterionId,
                tag = "Updated Tag",
                name = "Updated Criterion",
                description = "Updated Description",
                category = CriterionCategory.INCLUSION,
            )

            val updatedCriterion = repo.updateCriterion(request, emptySet())

            assertEquals(criterion, updatedCriterion)
        }
```

**Review by mowi12 — CHANGES_REQUESTED** [review]
Here is the first review. I will have time later to add additional notes.

Here some notes for you to think about for next time:

1. What about tests for the `fieldsFromGrpc` function? Something like:

```kotlin
class FromGrpcConversionTest {
    /**
     * The field enums are the single source of truth for the accepted field mask paths, so every path must actually
     * exist in the corresponding proto message.
     */
    @Nested
    inner class GrpcContract {
        @ParameterizedTest
        @EnumSource(UserField::class)
        fun `When a user field is checked, then its path exists in the proto message`(field: UserField) {
            assertThat(FieldMaskUtil.isValid(User.Update.getDescriptor(), field.grpcPath)).isTrue()
        }

        @ParameterizedTest
        @EnumSource(PaperField::class)
        fun `When a paper field is checked, then its path exists in the proto message`(field: PaperField) {
            assertThat(FieldMaskUtil.isValid(Paper.Update.getDescriptor(), field.grpcPath)).isTrue()
        }

        @ParameterizedTest
        @EnumSource(ProjectField::class)
        fun `When a project field is checked, then its path exists in the proto message`(field: ProjectField) {
            assertThat(FieldMaskUtil.isValid(Project.Update.getDescriptor(), field.grpcPath)).isTrue()
        }

        @ParameterizedTest
        @EnumSource(ProjectInfoField::class)
        fun `When a project info field is checked, then its path exists in the proto message`(field: ProjectInfoField) {
            assertThat(FieldMaskUtil.isValid(Project.Information.getDescriptor(), field.grpcPath)).isTrue()
        }

        @ParameterizedTest
        @EnumSource(CriterionField::class)
        fun `When a criterion field is checked, then its path exists in the proto message`(field: CriterionField) {
            assertThat(FieldMaskUtil.isValid(Criterion.Update.getDescriptor(), field.grpcPath)).isTrue()
        }
    }

...
```

This would be very exhaustive, but let's think about it.

2. I think we have some issues in the `ensureFieldMaskIsValid` function of the ValidationHelper file. The snowballr server calls `FieldMaskUtil.normalize` while the helper does not, so we do not check the same list of paths.

I think something like this would fix this:

```kotlin
/**
 * Ensures that the provided field mask is non-blank and contains only [allowedPaths].
 * If the field mask is not valid, a [InvalidFieldMask] validation issue is raised.
 *
 * The field enums, e.g. `UserField`, are the single source of truth for the acceptable paths of a request. Paths that
 * exist in the proto definition but have no matching field, such as read-only fields or the path of a message itself,
 * are rejected as well, because the layers below cannot act on them.
 *
 * @param fieldMask The [FieldMask] to validate.
 * @param allowedPaths The paths that may appear in the field mask.
 * @param allowEmpty Whether to allow an empty field mask. Defaults to false.
 */
fun Raise<ValidationIssue>.ensureFieldMaskIsValid(
    fieldMask: FieldMask,
    allowedPaths: Collection<String>,
    allowEmpty: Boolean = false,
) {
    val paths = FieldMaskUtil.normalize(fieldMask).pathsList

    ensure(allowEmpty || paths.isNotEmpty()) { InvalidFieldMask.createForBlankFieldMask() }

    val unknownPaths = paths.filterNot { it in allowedPaths }
    ensure(unknownPaths.isEmpty()) { InvalidFieldMask.createForContainsInvalidFields(unknownPaths) }
}
```

File: src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt (hunk 1 above)
Comment (Slartibartfass2) [reply src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt:134]: I thought I did that because I checked every update repo call (users, projects), but somehow I missed this or thought that it shouldn't be here... not sure.

I added it now.

File: src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt (hunk 3 above)
Comment (Slartibartfass2) [reply src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:295]: I added it.

File: src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt (hunk 2 above)
Comment (Slartibartfass2) [reply src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt]: Let's not overcomplicate this. This is a temporary conversion layer for a reason. I don't think we have to optimize this any further.

**Review by Slartibartfass2 — COMMENTED** [review]
Thanks for the review. Regarding your comments:
1. I'm fine with adding tests. Now that I moved the grpc paths to the grpc package (where they should belong), it's a bit difficult to test it. Do you have any suggestions? The main problem is that we cannot simply get all possible fields of a field mask afaik
2. I added the normalization, but what about the other changes? Where do they come from?

File: src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt
Comment (mowi12) [inline src/main/kotlin/se/uulm/snowballr/backend/grpc/FromGrpcConversion.kt]: review

File: src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt [hunk 4]
```diff
@@ -151,11 +151,13 @@ fun Raise<ValidationIssue>.ensureFieldMaskIsValid(
...
+    ensure(allowEmpty || paths.isNotEmpty()) { InvalidFieldMask.createForBlankFieldMask() }
     ensure(FieldMaskUtil.isValid(descriptor, fieldMask)) {
-        InvalidFieldMask.createForContainsInvalidFields(fieldMask.pathsList)
+        InvalidFieldMask.createForContainsInvalidFields(paths)
     }
-    ensure(fieldMask.pathsList.toSet().none { unallowedFields.contains(it) }) {
+    ensure(paths.none { unallowedFields.contains(it) }) {
         InvalidFieldMask.createForContainsUnallowedFields(unallowedFields)
     }
 }
```
Comment (mowi12) [inline src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt:159]: The mask that validation accepts and the mask the server can actually convert are two different sets, and the smaller one is the server's.

`FieldMaskUtil.isValid` accepts anything the proto descriptor knows, including read-only fields and bare message paths. I asked Claude to verify this and it did with real calls.

All calls passed validation and then hit `NoSuchElementException` in `paperFieldFromGrpc`. `ExceptionInterceptor` has no mapping for it, so the client gets `NOT_FOUND` for a bad argument.

The denylist cannot close hits, because it only names the fields we happened to think of. Every field added to the proto defaults to allowed, and failure mode is a crash rather than a rejection.

Inverting it to an allowlist fixes that and is less code:


```suggestion
fun Raise<ValidationIssue>.ensureFieldMaskIsValid(
    fieldMask: FieldMask,
    allowedPaths: Collection<String>,
    allowEmpty: Boolean = false,
) {
    val paths = FieldMaskUtil.normalize(fieldMask).pathsList

    ensure(allowEmpty || paths.isNotEmpty()) { InvalidFieldMask.createForBlankFieldMask() }

    val unknownPaths = paths.filterNot { it in allowedPaths }
    ensure(unknownPaths.isEmpty()) { InvalidFieldMask.createForContainsInvalidFields(unknownPaths) }
}
```

This will change the tests in `PaperValidatorTest` since they use `paper.id` as their deliberate unrelated mask path. That path is now rejected, so need a different value such as `paper.year`.

Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt:159]: This will also influence the validators e.g. the paperValidor:

```kotlin
    private const val FIELD_PAPER_AUTHORS = "paper.authors"
    private const val FIELD_PAPER_PUBLICATION_TYPE = "paper.publication_type"
    private const val FIELD_PAPER_PUBLICATION_NAME = "paper.publication_name"
    private const val FIELD_PAPER_PUBLISHER = "paper.publisher"
    private const val FIELD_PAPER_YEAR = "paper.year"
    private const val FIELD_PAPER_ABSTRACT = "paper.abstrakt"
    private const val FIELD_PAPER_TITLE = "paper.title"
    private const val FIELD_PAPER_EXTERNAL_IDS = "paper.external_ids"

    private val ALLOWED_UPDATE_PATHS = listOf(
        FIELD_PAPER_TITLE,
        FIELD_PAPER_ABSTRACT,
        FIELD_PAPER_YEAR,
        FIELD_PAPER_PUBLISHER,
        FIELD_PAPER_PUBLICATION_NAME,
        FIELD_PAPER_PUBLICATION_TYPE,
        FIELD_PAPER_AUTHORS,
        FIELD_PAPER_EXTERNAL_IDS,
    )
```

And then after:

```kotlin
ensureFieldMaskIsValid(request.mask, ALLOWED_UPDATE_PATHS)
```

**Review by mowi12 — CHANGES_REQUESTED** [review]
Your new commits look good. Taking `grpcPath` off the enums is a good call and better than putting them there in the first place. This keep the domain types clean for when  gRPC goes away.

One blocker before we can merge: validation and conversion still disagree about which mask paths are legal, so a client can send a proto-valid path such as `paper.backward_referenced_ids` and get `NOT_FOUND` and we get a warning in the console. Details are in `ValidationHelper.kt`.

One other thing worth nothing and we need to think about. You changed the behaviour for clients. Masks that used to be silently ignored will now be rejected. This is one of the reasons which I wanted to make the conversion more generic so we can improve it after the unifying instead of in 5 places. Your decision.

One last thing:
Yes, we will remove gRPC with Rest, but if we make mistakes now or omit testing and then replace untested behaviour with REST we cannot guarantee that we correctly replace it. We should still try to maintain quality in our code and not try to do it quickly so its working.

File: src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt (hunk 4 above)
Comment (Slartibartfass2) [reply src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt:159]: I think I got it now. I synced the validation and the extraction. Furthermore, I also used the frontend to verify this. It should be alright now. I see that this is not the best solution. We probably should've waited with this refactoring. Maybe, before we migrate, we should also try to sync both APIs as good as possible so that the frontend switch is not that big.

For example, we could now release a new API version in which we remove the field masks completely and only use full-replace updates.

Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt:159]: I'm fine with the idea if you want to do it that way.

**Review by mowi12 — APPROVED** [review]
